Topics/Tech_Stacks/WebSockets/Web-Sockets.md

```diff
@@ -0,0 +1,169 @@
+const WebSocket = require('ws');
+
+// Create a WebSocket server by instantiating a new WebSocket.Server object.
+// The server listens on port 8080 for incoming connections.
+const wss = new WebSocket.Server({ port: 8080 });
+
+// Set up an event listener for 'connection' events.
+// This event is emitted whenever a new client connects to the server.
+wss.on('connection', function connection(ws) {
+  
+  // Inside the connection event listener, set up another listener for 'message' events.
+  // The 'message' event is emitted when a message is received from a client.
+  ws.on('message', function incoming(message) {
+    
+    // Log the received message to the console.
+    console.log('received: %s', message);
+  });
+
+  // Send a message back to the client to acknowledge the connection.
+  ws.send('Hello, client! Message from the server...');
+});
+```
+In this setup, the server listens on port 8080. When a client connects, the server prints received messages to the console and sends a greeting message back to the client.
+
+### Client Setup
+The client initiates the connection to the WebSocket server and defines how to handle different types of events like opening the connection, receiving messages, errors, and closing the connection.
+
+```javascript
+// Create a new WebSocket instance, specifying the URL of the WebSocket server.
+var ws = new WebSocket('ws://localhost:8080');
+
+// Define the 'onopen' event handler.
+// This function is called when the WebSocket connection is successfully established.
+ws.onopen = function() {
+  
+  // Send a greeting message to the server.
+  ws.send('Hello! Message from the client...');
+};
+
+// Define the 'onmessage' event handler.
+// This function is called when a message from the server is received.
+ws.onmessage = function(event) {
+  
+  // Log the message from the server to the console.
+  console.log('Message from server:', event.data);
+};
+
+// Define the 'onerror' event handler.
+// This function is called when an error occurs with the WebSocket connection.
+ws.onerror = function(error) {
+  
+  // Log the error details to the console.
+  console.error('WebSocket Error:', error);
+};
+
+// Define the 'onclose' event handler.
+// This function is called when the WebSocket connection is closed.
+ws.onclose = function(event) {
+  
+  // Log the fact that the WebSocket connection has been closed.
+  console.log('WebSocket connection closed:', event);
+};
+```
+In the client code, a WebSocket object is created and connected to ws://localhost:8080. Event handlers are defined for various events. When the connection is opened, a message is sent to the server. Incoming messages are logged to the console, and any errors or connection closures are handled appropriately.
+
+Note: Ensure that the WebSocket server is running before attempting to connect with the client, as the client will not be able to establish a connection if the server is not listening for connections.
+
+## Real-World Use Cases
```

Comment:
Consider expanding on each example with a brief explanation of how WebSockets enhance the user experience in that specific context.

Reply:
Noted